src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforceStreamingSource.java

```diff
@@ -0,0 +1,183 @@
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.cdap.etl.api.PipelineConfigurer;
+import co.cask.cdap.etl.api.streaming.StreamingContext;
+import co.cask.cdap.etl.api.streaming.StreamingSource;
+import co.cask.hydrator.plugin.spark.ReferenceStreamingSource;
+import co.cask.hydrator.salesforce.SObjectDescriptor;
+import co.cask.hydrator.salesforce.SalesforceSchemaUtil;
+import co.cask.hydrator.salesforce.authenticator.Authenticator;
+import co.cask.hydrator.salesforce.authenticator.AuthenticatorCredentials;
+import com.sforce.soap.partner.PartnerConnection;
+import com.sforce.soap.partner.sobject.SObject;
+import com.sforce.ws.ConnectionException;
+import org.apache.spark.api.java.function.Function;
+import org.apache.spark.streaming.api.java.JavaDStream;
+import org.apache.spark.streaming.api.java.JavaStreamingContext;
+import org.json.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.Objects;
+import javax.ws.rs.Path;
+
+/**
+ * Returns records in realtime created by Salesforce. To achieve this Salesforce Streaming API is used.
+ * We use cometd server to subscribe to Salesforce Push Topics and receive realtime updates from there
+ *
+ */
+@Plugin(type = StreamingSource.PLUGIN_TYPE)
+@Name(SalesforceStreamingSource.NAME)
+@Description(SalesforceStreamingSource.DESCRIPTION)
+public class SalesforceStreamingSource extends ReferenceStreamingSource<StructuredRecord> {
+  private static final Logger LOG = LoggerFactory.getLogger(SalesforceStreamingSource.class);
+
+  static final String NAME = "SalesforceStreaming";
+  static final String DESCRIPTION = "Salesforce Streaming";
+  private SalesforceStreamingSourceConfig config;
+  private Schema schema;
+
+  public SalesforceStreamingSource(SalesforceStreamingSourceConfig config) {
+    super(config);
+    this.config = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    try {
+      config.validate(); // validate when macros are not substituted
+      config.ensurePushTopicExistAndWithCorrectFields(); // run when macros are not substituted
+
+      String query = config.getPushTopicQuery();
+
+      if (query != null && !query.isEmpty()
+        && !config.containsMacro(SalesforceStreamingSourceConfig.PROPERTY_PUSHTOPIC_QUERY)) {
+
+        Schema schema = SalesforceSchemaUtil.getSchema(config.getAuthenticatorCredentials(),
+                                                       SObjectDescriptor.fromQuery(query));
+        pipelineConfigurer.getStageConfigurer().setOutputSchema(schema);
+      }
+    } catch (ConnectionException e) {
+      throw new RuntimeException("There was issue communicating with Salesforce", e);
+    }
+  }
+
+  @Override
+  public JavaDStream<StructuredRecord> getStream(StreamingContext streamingContext) throws ConnectionException {
+    config.validate(); // validate when macros are substituted
```

Comment:
should we validate twice? Doesn't it call configurePipeline after macros are substituted?

Reply:
Yes configurePipeline runs with aliases not substituted. Just checked this.
And also I saw this approach of calling validate two times on other plugins as well.